jenkinsci/ec2-fleet-plugin
# Add support for fleets created through the CreateFleet API

Issue #, if available:
#396 

Description of changes:
Added the EC2CreateFleet class which provides support for EC2 Fleets created from the Create Fleet API.

Manual Testing:
- Created an EC2 Launch Template using an Amazon Linux 2 AMI.
- Created a JSON file, named fleet.json, for the Launch Template configs which specifies the launch template just created, the overrides which detail the weight capacity used to scale the number of executors, and target capacity specification used to specify the total weighted capacity of launched instances and the capacity type of the instances.
<img width="446" alt="Screenshot 2023-10-09 at 10 43 44 AM" src="https://github.com/jenkinsci/ec2-fleet-plugin/assets/66969321/d488fe80-27bf-4471-8097-e9cb84bfb0e1">

-  Ran "aws ec2 create-fleet --cli-input-json file://fleet.json" to create the EC2 Fleet
- Wait a couple minutes for the new instances to be launched
- Ran the Jenkins EC2-Plugin. Configured the cloud to use the newly created Fleet, set the number of executors to 1 and Scale Executors by Weight to false. This tests the instances unweighted. 
- Ran a test job to ensure the instances can run a job successfully.
<img width="340" alt="Screenshot 2023-10-09 at 10 44 43 AM" src="https://github.com/jenkinsci/ec2-fleet-plugin/assets/66969321/259d06cf-439c-4379-a168-de74333c68e2">

- Configure the cloud settings to enable Scale Executors by Weight. Restart the Jenkins EC2-Plugin to reconfigure the number of executors on the existing instances.
- Check that the number of executors have scaled appropriately. The instances ran are both c4.xlarge which have a weighted capacity of 2
<img width="342" alt="Screenshot 2023-10-09 at 10 47 53 AM" src="https://github.com/jenkinsci/ec2-fleet-plugin/assets/66969321/d8b0977c-79c9-495d-a260-1c65ba7b4939">




By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
